### ISSUE TITLE
shrinkwrap and optionalDependencies

### ISSUE THREAD

**Issue by seanmonstar** [npm/npm#2679]
We have to shrinkwrap, since some of our dependencies are loose, and that's no good for production. Still, some of the dependencies are optional, depending on the OS.

If we include a shrinkwrap file, then npm doesn't even try to install optional dependencies. And if we shrinkwrap after the optional has been installed, it's no longer optional.

**Comment by jumoel** [comment]
+1

This is still a problem.

**Comment by sarith** [comment]
+1

We just ran into this as well.
Any workarounds?

**Comment by othiym23** [comment]
This is now tagged onto the [npm road map](https://github.com/npm/npm/wiki/Roadmap#refinements-to-shrinkwrap-and-bundleddependencies). I can't promise that this is something we'll get to immediately, but the npm CLI team does plan to address this within the medium term (6-12 months). It will probably require making some hard choices about how to handle this case in a robust way, and may end up being a semver-major change (in that shrinkwraps that fix this may end up not being interoperable with older versions of npm).

Also, this is at least partially a duplicate of #6801.

**Comment by jackwanders** [comment]
Hi @othiym23, my organization has also been suffering from this issue. I've read through the v3 source involved in installing and shrink-wrapping dependencies, and from a technical angle, it seems feasible to create a shrinkwrap process that can account for the installation failure of optional dependencies.

The two scenarios that seem to require resolution are:

1. Properly allowing a shrinkwrapped optional dependency to fail on unsupported platforms
2. Properly installing an optional dependency of a shrinkwrapped dependency that itself was not included on the shrinkwrap (i.e. the shrinkwrap occurred on the unsupported platform)

For the first scenario, would it be possible to simply check if the package's parent has it listed in its `optionalDependencies`, and if so, adding an optional failure rollback to its `installSteps`?

The second scenario seems a bit more complicated. It would seem that, during the shrinkwrap process, you would want to iterate through all optional dependencies, generating an `optionalDependencies` block within `npm-shrinkwrap.json`, including data for all such packages, regardless of whether they are currently installed. Essentially, we want to provide as much data as possible about what WOULD have been installed on the system that's doing the shrinkwrap, assuming the installation did not fail. This would both generate exact specs of non-installed optional dependencies and allow future installations to attempt their own installation of these dependencies, rolling back on failure.

I'm curious to know if there is anything glaringly infeasible about this approach. Given the impact this is having on my team, I'm invested in at least discovering if a resolution is possible, and contributing to the cause if it is.

**Comment by lxe** [comment]
The workaround I've been doing was:

1. Remove the entry of your optional dependency from your `npm-shrinkwrap.json` tree
2. Add it as an "optionaDependency" of your shrinkwrapped project. 

Not ideal, but easy-ish to automate, and not too difficult to manage.

**Comment by mikemaccana** [comment]
Thanks @lxe I've been doing the same and it works.

**Comment by caseyhoward** [comment]
I'm sure a lot of people develop on an OS different from their build servers. How are you supposed to even use an npm-shrinkwrap file in this scenario?

**Comment by domharrington** [comment]
@caseyhoward i've been using the suggestion here: https://github.com/npm/npm/issues/2679#issuecomment-150084700. I don't think I even needed to remove it from the shrinkwrap.

**Comment by IanVS** [comment]
Here's my attempt at quick-and-dirty script to remove optional dependencies from a shrinkwrap file:

```js
'use strict';

var fs = require('fs');
var shrinkwrap = require('npm-shrinkwrap.json');
var optionalDeps = Object.keys(require('package.json').optionalDependencies);

optionalDeps.forEach(function(optionalDep) {
    delete shrinkwrap.dependencies[optionalDep];
});

fs.writeFileSync('npm-shrinkwrap.json', JSON.stringify(shrinkwrap, null, 2), 'utf8');
```

**Comment by riy** [comment]
@IanVS, I came up with a very similar (but not as elegant) solution just a few minutes ago.

After seeing your comment suddenly show up here while I had this tab open :-) I merged your removal code into mine and came up with a solution that replaces the need for `npm shrinkwrap` with `gulp shrinkwrap`, nicely removing the optional dependencies (e.g. those that shouldn't be installed because they're meant for a different architecture than that of your build server) right after shrinkwrapping.

Install the `gulp-shrinkwrap` plugin:
```
npm install gulp-shrinkwrap --save-dev
```

Specify a list of `optionalDependencies` in the `package.json`, like so:
```
"optionalDependencies": {
  "fsevents": "*"
},
```

And now add this to your Gulpfile:

```javascript
'use strict';

var fs = require('fs');
var shrinkwrap = require('gulp-shrinkwrap');

// The whole purpose of this task is to circumvent an issue with npm-shrinkwrap that doesn't take optional dependencies showing up in the npm-shrinkwrap.json into consideration
gulp.task('shrinkwrap', function() {
  // First we need to shrinkwrap as we usually do
  gulp.src('package.json') //
  .pipe(shrinkwrap()) // just like running 'npm shrinkwrap'
  .pipe(gulp.dest('./')) // save as 'npm-shrinkwrap.json'
  .on('end', function() {
    // Now load the npm-shrinkwrap.json
    var shrinkwrapJson = JSON.parse(fs.readFileSync('./npm-shrinkwrap.json'));
    // Remove optional dependencies (those that are listed in the package.json)
    var optionalDeps = Object.keys(require('./package.json').optionalDependencies);
    optionalDeps.forEach(function(optionalDep) {
      delete shrinkwrapJson.dependencies[optionalDep];
    });

    // And now overwrite the existing npm-shrinkwrap.json
    fs.writeFileSync('./npm-shrinkwrap.json', JSON.stringify(shrinkwrapJson, null, 2), 'utf8');
  })
});
```

So instead of `npm shrinkwrap` you now just run ` gulp shrinkwrap` and everything is fine. No optional dependency showing up in `npm-shrinkwrap.json` anymore, and the Linux build that once tried to load `fsevents`, because it was showing up in `npm-shrinkwrap.json`, is also fine.

Now if only some JavaScript pro could come up with a solution to place the removal code into a `.pipe()` between the existing two `.pipe()`-commands...

**Comment by glen-84** [comment]
@riy,

Maybe look at [gulp-modify](https://www.npmjs.com/package/gulp-modify) or [gulp-change](https://www.npmjs.com/package/gulp-change). Or do it [without a plug-in](http://stackoverflow.com/a/31272783/221528).

Your Gulp task should probably also return the stream.

**Comment by riy** [comment]
@glen-84

I'll try out [gulp-modify](https://www.npmjs.com/package/gulp-modify), thanks!

**Comment by riy** [comment]
Here's an update, based on @glen-84's feedback:

Install the `gulp-shrinkwrap` and `gulp-modify` plugins:
```
npm install gulp-shrinkwrap --save-dev
npm install gulp-modify --save-dev
```

Specify a list of `optionalDependencies` in the `package.json`, like so:
```
"optionalDependencies": {
  "fsevents": "*"
},
```

And now add this to your Gulpfile:

```javascript
'use strict';

var shrinkwrap = require('gulp-shrinkwrap');
var modify = require('gulp-modify');

// The whole purpose of this task is to circumvent an issue with npm-shrinkwrap that doesn't take optional dependencies showing up in the npm-shrinkwrap.json into consideration
gulp.task('shrinkwrap', function() {
  // First we need to shrinkwrap as we usually do
  return gulp.src('package.json') //
  .pipe(shrinkwrap()) // just like running 'npm shrinkwrap'
  .pipe(modify({
    fileModifier : function(file, content) {
      // Parse the string as JSON
      var shrinkwrappedJson = JSON.parse(content);
      // Remove optional dependencies (those that are listed in the package.json)
      var optionalDeps = Object.keys(require('./package.json').optionalDependencies);
      optionalDeps.forEach(function(optionalDep) {
        delete shrinkwrappedJson.dependencies[optionalDep];
      });

      return JSON.stringify(shrinkwrappedJson, null, 2);
    }
  })) //
  .pipe(gulp.dest('./')); // save as 'npm-shrinkwrap.json'
});
```

Instead of `npm shrinkwrap` you now just run ` gulp shrinkwrap` and everything is fine.

**Comment by henrytseng** [comment]
@riy Doesn't look like that works?  Getting the following:

```
$ gulp shrinkwrap
[02:42:05] Using gulpfile ~/.../gulpfile.js
[02:42:05] Starting 'shrinkwrap'...
npm WARN shrinkwrap Excluding devDependency: babel-eslint
npm WARN shrinkwrap Excluding devDependency: eslint
npm WARN shrinkwrap Excluding devDependency: gulp
npm WARN shrinkwrap Excluding devDependency: gulp-coverage
npm WARN shrinkwrap Excluding devDependency: gulp-develop-server
npm WARN shrinkwrap Excluding devDependency: gulp-eslint
npm WARN shrinkwrap Excluding devDependency: gulp-mocha
npm WARN shrinkwrap Excluding devDependency: gulp-plumber
npm WARN shrinkwrap Excluding devDependency: gulp-shrinkwrap
npm WARN shrinkwrap Excluding devDependency: gulp-watch
npm WARN shrinkwrap Excluding devDependency: mocha
npm WARN shrinkwrap Excluding devDependency: run-sequence
npm WARN shrinkwrap Excluding devDependency: supertest
wrote npm-shrinkwrap.json
Unhandled rejection Error: Cannot find module './package.json'
    at Function.Module._resolveFilename (module.js:339:15)
    at Function.Module._load (module.js:290:25)
    at Module.require (module.js:367:17)
    at require (internal/module.js:16:19)
    at gulp.src.pipe.pipe.modify.fileModifier (/Users/.../tasks/shrinkwrap.js:20:38)
    at DestroyableTransform._transform (/Users/.../node_modules/gulp-modify/index.js:22:23)
    at DestroyableTransform.Transform._read (/Users/.../node_modules/readable-stream/lib/_stream_transform.js:159:10)
    at DestroyableTransform.Transform._write (/Users/.../node_modules/readable-stream/lib/_stream_transform.js:147:83)
    at doWrite (/Users/.../node_modules/readable-stream/lib/_stream_writable.js:313:64)
    at writeOrBuffer (/Users/.../node_modules/readable-stream/lib/_stream_writable.js:302:5)
    at DestroyableTransform.Writable.write (/Users/.../node_modules/readable-stream/lib/_stream_writable.js:241:11)
    at write (/Users/.../node_modules/gulp-shrinkwrap/node_modules/through2/node_modules/readable-stream/lib/_stream_readable.js:623:24)
    at flow (/Users/.../node_modules/gulp-shrinkwrap/node_modules/through2/node_modules/readable-stream/lib/_stream_readable.js:632:7)
    at DestroyableTransform.pipeOnReadable (/Users/.../node_modules/gulp-shrinkwrap/node_modules/through2/node_modules/readable-stream/lib/_stream_readable.js:664:5)
    at emitNone (events.js:80:13)
    at DestroyableTransform.emit (events.js:179:7)
```

**Comment by henrytseng** [comment]
Here's the issue: our team breaks up gulp file tasks into subfolders since there a number of them.  Here's a quick fix with `process.cwd()`:

```
var gulp = require('gulp');
var shrinkwrap = require('gulp-shrinkwrap');
var modify = require('gulp-modify');

// The whole purpose of this task is to circumvent an issue with npm-shrinkwrap that doesn't take optional dependencies showing up in the npm-shrinkwrap.json into consideration
gulp.task('shrinkwrap', function() {
  return gulp.src('package.json') //
  .pipe(shrinkwrap()) // just like running 'npm shrinkwrap'
  .pipe(modify({
    fileModifier : function(file, content) {
      var shrinkwrappedJson = JSON.parse(content);
      var optionalDeps = Object.keys(require(process.cwd()+'/package.json').optionalDependencies);
      optionalDeps.forEach(function(optionalDep) {
        delete shrinkwrappedJson.dependencies[optionalDep];
      });

      return JSON.stringify(shrinkwrappedJson, null, 2);
    }
  }))
  .pipe(gulp.dest('./'));
});
```

**Comment by riy** [comment]
@henrytseng I have to admit I'm not that familiar with Gulp tasks but I'm happy you found a solution that works for you.

**Comment by tatablack** [comment]
Since this is still open, inspired by @riy's [comment](https://github.com/npm/npm/issues/2679#issuecomment-219741669) I wrote a simple grunt task that generates a `npm-shrinkwrap.json` file and then removes from it a list of dependencies you specify.

Code's here: https://github.com/Skyscanner/grunt-custom-shrinkwrap, I'm happy to take suggestions or PRs.

**Comment by palmerj3** [comment]
Hi all have a look at this PR. I think the best approach here is to remove optionalDependencies from generated shrinkwrap files.  https://github.com/npm/npm/pull/13579

**Comment by rymohr** [comment]
What if `npm install` simply warned on `EBADPLATFORM` instead of erroring when installing from shrinkwrap?

**Comment by rymohr** [comment]
Sorry, just saw https://github.com/npm/npm/issues/11632.

**Comment by JamieMason** [comment]
[shrinkpack](https://github.com/JamieMason/shrinkpack) might be able to help those of you who've been trying to check in your dependencies and have been having cross-platform issues. It will also remove `optionalDependencies` from `npm-shrinkwrap.json` for you by default.

**Comment by caseyhoward** [comment]
If this issue isn't going to get fixed, can we get something added to the caveats section of the documentation (https://docs.npmjs.com/cli/shrinkwrap#caveats) that npm shrinkwrap isn't meant to be used between different environments? The optionalDependencies documentation should also be updated to say that it's not compatible with "npm shrinkwrap". It could save people a lot of time. I don't know how many this issue has bitten me.

**Comment by Kovensky** [comment]
I really wish this wasn't just closed as "won't fix, the bug is now intended behavior".

Documenting it as a caveat is fine, but not that it is not _meant_ to be used between different environments, just that it isn't currently supported.

At least I'm pretty sure the overwhelming majority of Mac developers don't use Mac servers for their production hosts. You would also have to get all your developers to run the same OS. Systems like Vagrant and Docker help, but requiring developing in a VM to be able to deal with shrinkwraps is not very nice.

**Comment by palmerj3** [comment]
@JamieMason apparently [this makes no difference](https://github.com/npm/npm/pull/13579#issuecomment-243898968), i've learned.

**Comment by sean-krail** [comment]
@IanVS, I modified [your script](https://github.com/npm/npm/issues/2679#issuecomment-219103907) and my `package.json` file a bit to have it run after `npm shrinkwrap` with node v6.10.0 and npm v3.10.10:
```JavaScript
"use strict";

// https://github.com/npm/npm/issues/2679#issuecomment-219103907

let fs = require("fs");

let packageJson = fs.readFileSync("./package.json");
if (!packageJson) return "No package.json";

let shrinkwrap = fs.readFileSync("./npm-shrinkwrap.json");
if (!shrinkwrap) return "No npm-shrinkwrap.json";

shrinkwrap = JSON.parse(fs.readFileSync("./npm-shrinkwrap.json"));

let optionalDeps = JSON.parse(packageJson).optionalDependencies;
if (!optionalDeps) return "No optional dependencies in package.json";

optionalDeps = Object.keys(optionalDeps);
optionalDeps.forEach(optionalDep => delete shrinkwrap.dependencies[optionalDep]);

fs.writeFileSync("./npm-shrinkwrap.json", JSON.stringify(shrinkwrap, null, 2), "utf8");
```

With a `package.json` file like this:
```JSON
{
  "scripts": {
    "postshrinkwrap": "node {{path to file above}}"
  },
  "dependencies": {
    "...": "..."
  },
  "devDependencies": {
    "...": "..."
  },
  "optionalDependencies": {
    "fsevents": "*"
  },
  "...": "..."
}
```

**Comment by felixfbecker** [comment]
This should be a way higher priority to fix. yarn has shown that people need lockfiles, and npm's lockfile system (shrinkwrap) is obviously broken

**Comment by thom-nic** [comment]
FWIW, yarn has also had issues with optional dependencies:
* yarnpkg/yarn#1059 (marked closed but users are still reporting issues)
* yarnpkg/yarn#705
* yarnpkg/yarn#1233

For my part, I have had good luck with `--production --no-optional` using npm@4.2.0.

**Comment by ntwb** [comment]
Per the release of npm v5: http://blog.npmjs.org/post/161081169345/v500
> _"package locks no longer exclude optionalDependencies that failed to build. This means package-lock.json and npm-shrinkwrap.json should now be cross-platform. (#15900)"_

Does npm v5 fix this optionalDependencies issues, it sounds like it to me 🤔

**Comment by ntwb** [comment]
Arghhhh... I just created a `package-lock.json` on Mac where [fsevents](https://www.npmjs.com/package/fsevents) was installed and included in the `package-lock.json` file, I then created a PR and Travis CI failed the build.
 
I immediately thought of this issue which I thought was resolved as part of npm 5 (per my own above comment FWIW), turns out it's not.

Once I manually removed `fsevents` from `package-lock.json` Travis CI was happy once more 😢 

Here's the PR where this occurred: https://github.com/stylelint/stylelint/pull/2670

I'm going to close that PR and add `package-lock.json` to our `.gitignore` until this issue is sorted once and for all.
